### PR TITLE
Financial Connections: implemented retrieve auth method to account for partner auth redirect edge-cases

### PR DESCRIPTION
**Pull request by kgaidis-stripe**
## Summary

This PR implements a new "retrieve auth session" method that helps us fetch the current user auth session, and check for its success status, even if we did not get back a redirect from banks/providers.

Android PR: https://github.com/stripe/stripe-android/pull/7125

## Testing

Not all flows are easy to illustrate, but:
- Tested happy path oauth
- Tested happy path legacy
- Tested cancel oauth (a 'status' cancel and a native-cancel)
- Tested cancel legacy
- Tested app to app interrupt (see video below)
- Tested app to app non-interrupted (a successful flow, but I had an error)

Ran end to end tests (`bitrise run financial-connections-stability-tests`):

```
FinancialConnectionsNetworkingUITests
    ✓ testNativeNetworkingTestMode (73.879 seconds)
FinancialConnectionsUITests
    ✓ testDataLiveModeOAuthNativeAuthFlow (30.194 seconds)
    ✓ testDataLiveModeOAuthWebAuthFlow (25.594 seconds)
    ✓ testDataTestModeOAuthNativeAuthFlow (21.752 seconds)
    ✓ testPaymentTestModeLegacyNativeAuthFlow (24.329 seconds)
    ✓ testPaymentTestModeManualEntryNativeAuthFlow (28.183 seconds)
    ✓ testSearchInLiveModeNativeAuthFlow (25.912 seconds)

```

This video shows me interrupting Chase App To App flow, and even though I interrupted it (by not allowing it to go to Safari browser), it proceeded to check the auth session, which then resulted in proceeding to account picker (which then manifested into an error):

https://github.com/stripe/stripe-ios/assets/105514761/b3346e0d-dfd8-47cc-8afc-0f46aaa65382

### CODE REVIEW

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthDataSource.swift [hunk 1]
```diff
@@ -32,6 +34,9 @@ final class PartnerAuthDataSourceImplementation: PartnerAuthDataSource {
     private let clientSecret: String
     let analyticsClient: FinancialConnectionsAnalyticsClient
     let reduceManualEntryProminenceInErrors: Bool
+    var disableAuthSessionRetrieval: Bool {
+        return manifest.features?["bank_connections_disable_defensive_auth_session_retrieval_on_complete"] == true
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthDataSource.swift:38]: [Grep to double-check spelling](https://livegrep.corp.stripe.com/search/stripe?q=bank_connections_disable_defensive_auth_session_retrieval_on_complete&fold_case=auto&regex=false&context=true)

Note that this should ONLY return `true` if backend has this set as a flag. NULL == enabled.

I also tested this by setting the manifest manually in init.

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift [hunk 2]
```diff
@@ -629,6 +659,83 @@ final class PartnerAuthViewController: UIViewController {
...
         )
     }
+
+    // There are edge-cases where redirect links don't work properly.
+    // Check the auth session in-case the auth session was successful.
+    private func checkIfAuthSessionWasSuccessful(
+        authSession: FinancialConnectionsAuthSession,
+        completionHandler: @escaping (_ isSuccess: Bool) -> Void
+    ) {
+        guard !dataSource.disableAuthSessionRetrieval else {
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift:669]: all of this new logic should only execute if NOT disabled

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift [hunk 3]
```diff
@@ -113,10 +113,19 @@ final class PartnerAuthViewController: UIViewController {
                     self?.didSelectURLInTextFromBackend(url)
                 },
                 didSelectContinue: { [weak self] in
+                    guard let self = self else { return }
+                    self.dataSource.analyticsClient.log(
+                        eventName: "click.continue",
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift:118]: [Splunk](https://splunk.corp.stripe.com/en-US/app/search/search?sid=1691446979.383879_16A88E4A-C308-4A39-B80F-1C1910B5BF26)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift [hunk 4]
```diff
@@ -113,10 +113,19 @@ final class PartnerAuthViewController: UIViewController {
                     self?.didSelectURLInTextFromBackend(url)
                 },
                 didSelectContinue: { [weak self] in
+                    guard let self = self else { return }
+                    self.dataSource.analyticsClient.log(
+                        eventName: "click.prepane.continue",
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift:118]: [Splunk](https://splunk.corp.stripe.com/en-US/app/search/search?sid=1691447643.384057_16A88E4A-C308-4A39-B80F-1C1910B5BF26)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift [hunk 5]
```diff
@@ -408,6 +405,10 @@ final class PartnerAuthViewController: UIViewController {
             institutionImageUrl: self.institution.icon?.default,
             didSelectContinue: { [weak self] in
                 guard let self = self else { return }
+                self.dataSource.analyticsClient.log(
+                    eventName: "click.apptoapp.continue",
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift:409]: [Splunk](https://splunk.corp.stripe.com/en-US/app/search/search?sid=1691447736.384081_16A88E4A-C308-4A39-B80F-1C1910B5BF26)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift [hunk 6]
```diff
@@ -629,6 +659,83 @@ final class PartnerAuthViewController: UIViewController {
...
+
+    private func logUrlReceived(
+        _ url: URL?,
+        status: String?,
+        authSessionId: String
+    ) {
+        dataSource
+            .analyticsClient
+            .log(
+                eventName: "auth_session.url_received",
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift:730]: [Splunk](https://splunk.corp.stripe.com/en-US/app/search/search?sid=1691447896.384125_16A88E4A-C308-4A39-B80F-1C1910B5BF26)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift [hunk 7]
```diff
@@ -730,17 +846,36 @@ private extension PartnerAuthViewController {
...
              If we then proceed with authorization network request might fail as we will be doing background networking without special permission..
              */
             return
         }
-        handleAuthSessionCompletionFromNativeRedirect(url)
-        clearStateAndUnsubscribeFromNotifications()
+        if let url = unprocessedReturnURL {
+            if let authSession = dataSource.pendingAuthSession {
+                dataSource.recordAuthSessionEvent(
+                    eventName: "native-app-to-app-redirect-url-received",
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthViewController.swift:859]: [some example auth session events](https://admin.corp.stripe.com/bank-connections/auth-sessions/bcsess_1Ncc3PLRlMYnb6tuNoB7DyqB)

File: StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthDataSource.swift (hunk 1 above)
Comment (kgaidis-stripe) [reply StripeFinancialConnections/StripeFinancialConnections/Source/Native/PartnerAuth/PartnerAuthDataSource.swift:38]: [PR for feature](https://git.corp.stripe.com/stripe-internal/pay-server/pull/666331)

File: StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/FinancialConnectionsAPIClient.swift [hunk 8]
```diff
@@ -228,6 +233,17 @@ extension STPAPIClient: FinancialConnectionsAPIClient {
...
 
+    func retrieveAuthSession(
+        clientSecret: String,
+        authSessionId: String
+    ) -> Future<FinancialConnectionsAuthSession> {
+        let body = [
+            "client_secret": clientSecret,
+            "id": authSessionId,
+        ]
+        return self.post(resource: APIEndpointAuthSessionsRetrieve, object: body)
```
Comment (kgaidis-stripe) [inline StripeFinancialConnections/StripeFinancialConnections/Source/API Bindings/FinancialConnectionsAPIClient.swift:244]: [PR for API](https://git.corp.stripe.com/stripe-internal/pay-server/pull/666319)

**Review by ccen-stripe — APPROVED** [review]
LGTM, Not sure if the changes are covered by existing tests, it seems only impacted tests are just making sure it builds - will leave up to you to update

**Pull request merged**